Update to Major Version 3.1. Removes MXNet

fhieber opened the pull request:
Models trained with PyTorch in version `3.0.x` are compatible with 3.1.


#### Pull Request Checklist ##
- [x] Changes are complete (if posting work-in-progress code, prefix your pull request title with '[WIP]'
until you can check this box.
- [x] Unit tests pass (`pytest`)
- [x] Were system tests modified? If so did you run these at least 5 times to account for the variation across runs?
- [x] System tests pass (`pytest test/system`)
- [x] Passed code style checking (`./style-check.sh`)
- [x] You have considered writing a test
- [x] Updated major/minor version in `sockeye/__init__.py`. Major version bump if this is a backwards incompatible change.
- [x] Updated CHANGELOG.md
- [x] Create updated backwards compatibility model for testing


By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

fhieber commented:
Overall diff includes file renames. Best to review individual commits.

mjdenkowski commented:
Is it feasible to use a 2 step process for this PR?

1. Remove all MXNet code/files without renaming PyTorch files.

2. Rename the PyTorch files.

If we review after each checkpoint, we can isolate all of the large rename diffs into rename-only commits.

fhieber commented:
It was a lot easier to do this module by module to ensure things are working properly. I added a temporary commit, `af349d802a81b`  that renames back to `_pt.py` but keeps the updated class names and imports. This obviously doesn't work but gives a better diff. Once we are happy with that, we can revert `af349d802a81b`.

mjdenkowski reviewed (commented):
Looks good so far!

There are a few minor comments and a main question about handling/testing backward compatibility.

fhieber commented on sockeye/model_pt.py:
```diff
@@ -131,32 +128,23 @@ def __init__(self,
 
         # traced components (for inference)
         self.traced_embedding_source = None  # type: Optional[pt.jit.ScriptModule]
+        self.traced_embedding_target = None  # type: Optional[pt.jit.ScriptModule]
         self.traced_encoder = None  # type: Optional[pt.jit.ScriptModule]
+        self.traced_decoder = None  # type: Optional[pt.jit.ScriptModule]
+        self.traced_factor_output_layers = None  # type: Optional[List[pt.jit.ScriptModule]]
```
They must have crept in when merging from main. I'll remove.

fhieber commented on sockeye/optimizers.py:
```diff
@@ -65,7 +47,8 @@ class PyTorchOptimizerConfig(config.Config):
 
 
-def get_optimizer(model: torch.nn.Module, config: PyTorchOptimizerConfig) -> Tuple[torch.optim.Optimizer, Dict[str, Any]]:
+def get_optimizer(model: torch.nn.Module, config: OptimizerConfig) -> Tuple[torch.optim.Optimizer,
+                                                                            Dict[str, Any]]:
```
Yes

fhieber commented on sockeye/test_utils.py:
```diff
@@ -204,23 +207,11 @@ def run_train_translate(train_params: str,
-        prepare_data_mod = sockeye.prepare_data
-        train_mod = sockeye.train
-        translate_mod = sockeye.translate
+    prepare_data_mod = sockeye.prepare_data
+    train_mod = sockeye.train
+    translate_mod = sockeye.translate
```
good catch

fhieber commented on test/integration/test_backwards_compatibility.py:
```diff
@@ -56,25 +55,3 @@ def test_backwards_compatibility():
             assert model_out.read() == EXPECTED_OUTPUT
-
-
-def test_backwards_compatibility_pt():
```
Yes, my plan was to create a 3.0.x model for this test once we have the rest of the PR finished. I'll add a TODO item to the PR to not forget about this. We can also have 2 tests: one for 3.0 and one for 2.3 backwards compatibility

fhieber commented on sockeye/utils.py:
```diff
@@ -58,7 +52,7 @@ def check_version(version: str):
+    # TODO(fhieber): maintain backwards compatibility for 2.3.x models
     if given_version[0] == '2' and given_version[1] == '3':
         logger.info(f"Code version: {__version__}")
         logger.warning(f"Given release version ({version}) does not match code version ({__version__}). "
```
Good question. Technically people could have converted their 2.3.x models to PyTorch with Sockeye 3.0.x, and these still run with Sockeye 3.1.x. If a user tries to run with an MXNet 2.3.x model it silently do any bad things but will fail so I think this would be safe. 
What do you think?

mjdenkowski replied:
We could say that Sockeye 3.0 is the intermediate version that works with MXNet, PyTorch, and 2.3.x models.  Sockeye 3.1 is PyTorch-only and requires 3.0+ models.

This gives users the opportunity to convert models and run MXNet/PyTorch comparisons with 3.0.  The new PyTorch models they train will be compatible with 3.1.

fhieber replied:
sounds good. I'll make this change

fhieber commented on test/unit/test_beam_search.py:
```diff
@@ -18,179 +18,62 @@
-    from mxnet import np
-    import sockeye.beam_search
-
+def test_pytorch_candidate_scorer():
```
yes

fhieber commented:
@mjdenkowski: addressed all comments. If you are fine with the diff, we can rename back `_pt` files to make the tests pass.

mjdenkowski commented on README.md:
```diff
@@ -12,9 +12,14 @@ For a quickstart guide to training a standard NMT model on any size of data, see
-MXNet models in version 2.3.x a little bit longer. If MXNet 2.x is installed, Sockeye can run both with PyTorch or MXNet.
+### Version 3.1.x: PyTorch only
+With version 3.1.x, we remove support for MXNet 2.x. Models trained with PyTorch and Sockeye 3.0.x remain compatible
+with Sockeye 3.1.x. Models trained with 2.3.x (using MXNet) and converted to PyTorch with Sockeye 3.0.x' conversion
```
`Sockeye 3.0.x'` -> `Sockeye 3.0.x's`

fhieber replied:
good catch

mjdenkowski commented on README.md:
```diff
@@ -52,6 +57,7 @@ For faster GPU training, install [NVIDIA Apex](https://github.com/NVIDIA/apex).
 
 ### Older versions
 
+- Sockeye 3.0, based on PyTorch & MXNet 2.x is available in the `sockeye_30` branch.
```
Good idea keeping a branch with MXNet/2.3.x compatibility!

mjdenkowski reviewed (commented):
Thanks for iterating!  There's just one minor comment and then everything looks good to rename.

fhieber commented:
Alright
- I addressed the remaining comments and fixed the MANIFEST.in
- created a `sockeye_30` branch
- released latest 3.0.15 version to PyPi

If you approve, I can update the required Github actions for merging PRs to main to pick up the workflows run in this PR (without `mx-{yes,no}`) and then we can merge.

mjdenkowski commented:
Everything looks good to update the workflow.
